### ISSUE TITLE
Proposed custom build for SageMath

### ISSUE THREAD

**Issue by paulmasson** [paulmasson/threejs-sage#1]
@mkoeppe here is my proposed custom build of Three.js for use with SageMath. As you can see, the repository is very simple: just the build instructions, Three.js license and the build. I've changed the instructions relative to the [discussion](https://trac.sagemath.org/ticket/30620#comment:10) on Trac to key off GitHub tags rather than the `master` branch. This way anyone can follow the instructions and recreate the build at any future time to confirm the accuracy of the process.

This repository is my preferred way of addressing the the issue identified in [Trac #30620](https://trac.sagemath.org/ticket/30620). If we work from the full Three.js repository, we will force users to download a thousand times more data, most of which will be useless to them. If you concur, then this repository can be transferred to SageMath and I will commit to performing the build once a month for each new release.

@jcamp0x2a since Mr.doob [acknowledged](https://github.com/mrdoob/three.js/issues/20387#issuecomment-699561027) that he has no control of how the revision string appears in the final build, I've added a simple text file to the repository that will be updated for each release. Could you please test my builds against your ongoing updates to make sure I've included everything you need? Thanks!

**Comment by jcamp0x2a** [comment]
Hi Paul. Thanks for putting this together. I was able to replicate the build using the instructions provided, and I tested it out on my "fat" lines branch without issue. I've pushed an update to [Trac #26410](https://trac.sagemath.org/ticket/26410#comment:18) with my latest code. Would of course want to change the repository/CDN path when the repo gets moved.

**Comment by paulmasson** [comment]
Adding a couple more voices to the conversation. @egourgoulhon and @slel, would both of you be comfortable with the custom build process and final product as presented here? Are the any other people who would have a strong opinion on this?

**Comment by mkoeppe** [comment]
What is the final product? The minified .js file, or GitHub's automatic tarball archive of this repository, or ...? 

How will it be installed in Sage-the-distribution? How would a user install it who wishes to run it in a system jupyter notebook or JupyterLab? as discussed in https://trac.sagemath.org/ticket/30124, https://trac.sagemath.org/ticket/30123 ?

**Comment by paulmasson** [comment]
> 
> 
> What is the final product? The minified .js file, or GitHub's automatic tarball archive of this repository, or ...?

Since we are already building a tarball with the existing script, GitHub's tarball is the natural final product.

> How will it be installed in Sage-the-distribution? How would a user install it who wishes to run it in a system jupyter notebook or JupyterLab? as discussed in https://trac.sagemath.org/ticket/30124, https://trac.sagemath.org/ticket/30123 ?

For use in Sage, nothing needs to change right now: we're just copying the unzipped tarball to `local/share/threejs`. This product is not intended to be installed apart from Sage, so I'm not sure I understand what you are planning for the other situations.

**Comment by slel** [comment]
> Adding a couple more voices to the conversation. @egourgoulhon and @slel, would
> both of you be comfortable with the custom build process and final product as
> presented here? Are the any other people who would have a strong opinion on this?

I lack the skills to comment. Maybe post to sage-packaging to get voices of people
who package Sage in various ways?

**Comment by mkoeppe** [comment]
> > What is the final product? The minified .js file, or GitHub's automatic tarball archive of this repository, or ...?
> 
> Since we are already building a tarball with the existing script, GitHub's tarball is the natural final product.

I see, so `build/pkgs/threejs/spkg-src` would be removed, and GitHub's tarball would be the "upstream tarball". 
 
> > How will it be installed in Sage-the-distribution? How would a user install it who wishes to run it in a system jupyter notebook or JupyterLab? as discussed in https://trac.sagemath.org/ticket/30124, https://trac.sagemath.org/ticket/30123 ?
> 
> For use in Sage, nothing needs to change right now: we're just copying the unzipped tarball to `local/share/threejs`.

OK, I understand.

>  This product is not intended to be installed apart from Sage, so I'm not sure I understand what you are planning for the other situations.

One case is the following. We hope to make it possible to run the Sage jupyter kernel with a system jupyter notebook, see https://trac.sagemath.org/ticket/30476 

In fact, in https://trac.sagemath.org/attachment/ticket/30476/Essai%20interact-2.log you can see what is happening in this situation:
```
[W 17:58:50.776 NotebookApp] 404 GET /nbextensions/threejs/build/three.min.js (::1) 13.93ms referer=http://localhost:8888/notebooks/Temporaire/Sagemath/Essai%20interact.ipynb
[W 17:58:50.780 NotebookApp] 404 GET /nbextensions/threejs/examples/js/controls/OrbitControls.js (::1) 3.10ms referer=http://localhost:8888/notebooks/Temporaire/Sagemath/Essai%20interact.ipynb
```

**Comment by paulmasson** [comment]
> > > What is the final product? The minified .js file, or GitHub's automatic tarball archive of this repository, or ...?
> > 
> > 
> > Since we are already building a tarball with the existing script, GitHub's tarball is the natural final product.
> 
> I see, so `build/pkgs/threejs/spkg-src` would be removed, and GitHub's tarball would be the "upstream tarball".
> 
How does that work these days? Do I alter the version file directly for an upgrade? How does the checksum file get updated? How does the tarball make it to the mirrors?

**Comment by mkoeppe** [comment]
> > I see, so `build/pkgs/threejs/spkg-src` would be removed, and GitHub's tarball would be the "upstream tarball".
> 
> How does that work these days? Do I alter the version file directly for an upgrade? How does the checksum file get updated? How does the tarball make it to the mirrors?

You update it in a trac ticket. Add an `upstream_url` field to checksums.ini - as explained in https://wiki.sagemath.org/ReleaseTours/sage-9.1#For_developers-1
When the ticket is merged, the release manager uploads the file pointed to by the URL to the mirrors.

**Comment by paulmasson** [comment]
> > > I see, so `build/pkgs/threejs/spkg-src` would be removed, and GitHub's tarball would be the "upstream tarball".
> > 
> > 
> > How does that work these days? Do I alter the version file directly for an upgrade? How does the checksum file get updated? How does the tarball make it to the mirrors?
> 
> You update it in a trac ticket. Add an `upstream_url` field to checksums.ini - as explained in https://wiki.sagemath.org/ReleaseTours/sage-9.1#For_developers-1
> When the ticket is merged, the release manager uploads the file pointed to by the URL to the mirrors.

@mkoeppe so the procedure forward is

1) Add the URL template to checksums.ini
2) Upgrade the package on my local machine
3) Commit the changed files on a ticket

Right? And in the future, only the last two steps for each upgrade.

Is this information in the official documentation yet?

**Comment by mkoeppe** [comment]
That's right.
See https://doc.sagemath.org/html/en/developer/packaging.html

**Comment by paulmasson** [comment]
@mkoeppe the upstream_url is not described in the official documentation, and neither is the option to upgrade a package from the Sage command line.

More importantly, @jcamp0x2a has pending upgrades to the Three.js viewer that need an updated version of the package. Are you ready to have me transfer this repository to SageMath so that we can proceed?

**Comment by mkoeppe** [comment]
You are right - https://trac.sagemath.org/ticket/30895 is about adding this to the documentation.

**Comment by mkoeppe** [comment]
> @jcamp0x2a has pending upgrades to the Three.js viewer that need an updated version of the package. Are you ready to have me transfer this repository to SageMath so that we can proceed?

I don't really think that any upgrades are dependent on the location of the repository.

Before we add a repository to the sagemath organization on github, I would prefer it if the open questions regarding deployment are addressed. But this certainly does not block any of the upgrades that you have in mind.
